DO NOT MERGE.  Attempt to move instrumentation over to Istanbul.

sam-falvo opened the pull request:
A baby-step heading towards integrating Istanbul as envisioned by Tomaz.  The mapping from jscoverage to Istanbul is far from perfect, as this early work already demonstrates.  Maybe there's a better way?

I'm an advocate of using istanbul in its native way, but that idea has already been dismissed as impractical or incompatible with existing whiskey users.

Commit:
Attempt to move instrumentation over to Istanbul.

Kami commented on lib/run.js:
```diff
@@ -517,30 +517,32 @@ function run(cwd, argv) {
-              err = new Error('jscoverage binary not found. To use test coverage ' +
-                              ' you need to install node-jscoverag binary - ' +
-                              'https://github.com/visionmedia/node-jscoverage');
+            if (err.message.match(/istanbul: not found/i)) {
```
I'm fine with adding `istanbul` package to dependencies in `package.json`.

We didn't add jscoverage, because it was pain to install it and requires C++ module to be compiled.

robert-chiniquy replied:
Sweet

Kami replied:
What I said in my previous comment - we can get rid of this stuff with istanbul.

sam-falvo replied:
I'm leaving this code in for now, but I've updated the package.json file to list Istanbul as a dependency.

Kami replied:
Lets remove unnecessary code and change it to exec `node_modules/.bin/istanbul` otherwise it won't really work if you don't install istanbul package globally (`npm install instanbul -g`)

sam-falvo replied:
Ahh!  Good catch and good point.  Fixed in 9578824df0af42cc3dded93535c757e5faa1209d

Commit:
Fixed missing istanbul instrumentation command

Commit:
blah

Commit:
Code cleanup prepping for todays editing.

Commit:
Work done to convert Istanbul coverage objects into jsCoverage objects.

The conversion seems to work, at least from a type-safety point of view.
However, both HTML and CLI reporters indicate 0 lines of code covered.
Using the json reporter, however, clearly indicates otherwise.

To run:
NODE_PATH=$(pwd)/lib-cov bin/whiskey --coverage --coverage-reporter cli --coverage-file /tmp/hooey --coverage-dir /tmp -t example/test-bdd.js

Commit:
typo

Commit:
Coverage reports for CLI reporter working again

robert-chiniquy commented on example/test-bdd.js:
```diff
@@ -15,7 +15,7 @@
  */
 
-var bdd = require('../lib/bdd').init(exports);
+var bdd = require('../lib-cov/bdd').init(exports);
```
Mentioning again: https://github.com/jhnns/rewire

sam-falvo replied:
I agree; but for now I'd like to adopt rewire in a separate PR, to keep this PR focused on Istanbul integration.  Hand-changing test-bdd.js like this is just a "get the thing working" patch that I'll revert when I'm happy everything else works as I expected.

robert-chiniquy replied:
I think it would be nice to have a test case for this which also demonstrates how to use whiskey with istanbul.

robert-chiniquy replied:
Like in `tests/run.sh` or something. Since istanbul is going into the dependencies for whiskey this should be fine without adding a vm.

Commit:
HTML reporter once again has proper source listing.

robert-chiniquy commented on lib/run_test_file.js:
```diff
@@ -23,6 +23,7 @@ var common = require('./common');
 var testUtil = require('./util');
 var coverage = require('./coverage');
 var constants = require('./constants');
+var fs = require("fs");
```
unused?

sam-falvo replied:
Unused, but relatively low priority for me at the moment.  I'll remove it in a separate clean-up pass.  My big concern now is, am I doing things the right way?  For example, the HTML reporter indicates a lot of lines hit, yet it says 25 out of 29 lines missed.  It doesn't make much sense to me yet.

Commit:
Fixed abnormally low hit-vs-missing bug in coverage reports

Kami commented on lib/coverage.js:
```diff
@@ -22,15 +22,46 @@
  */
+
+function jscoverageFromIstanbulCoverage(coverage) {
+  jscov = {};
```
Missing var

Kami commented on lib/coverage.js:
```diff
@@ -22,15 +22,46 @@
+  jscov = {};
+
+  for (var file in coverage) {
+    filename = path.basename(file);
```
We should use absolute paths or paths relative to the project root directory.

Otherwise it will really only work with projects which don't have nested directories and have unique file names.

sam-falvo replied:
Let's make this a separate PR, so that I can keep things focused.  jscoverage used relative pathnames, so for the time being, I'm preserving its behavior.  I'd like to address this next.

Kami replied:
I'm pretty sure it uses paths relative to the project directory (e.g. `lib/messenger/server.js`) otherwise it would never work with nested directories.

Kami commented on lib/coverage.js:
```diff
@@ -22,15 +22,46 @@
+    filename = path.basename(file);
+    jscov[filename] = {};
+
+    jscov[filename]["source"] = fs.readFileSync(file).toString().split("\n");
```
Style thing - single quotes around strings.

Kami commented on lib/coverage.js:
```diff
@@ -76,7 +107,9 @@ function coverage(data, type) {
   }
 
-  for (var i = 0, len = data.lines.length; i < len; ++i) {
+  len = Object.keys(data.lines).length;
```
Missing var (len is leaking into global scope).

Lets try to get this branch merged first and then I will add jshint file and clean up all the style and consistency issues with defining variables at the beginning of a function.

Kami commented on lib/run.js:
```diff
@@ -517,30 +517,32 @@ function run(cwd, argv) {
+      // }
 
-      coverageArgs.push(sprintf('lib %s', constants.COVERAGE_PATH));
+      coverageArgs.push(sprintf('--output %s lib', constants.COVERAGE_PATH));
```
Instead of shelling out, can we actually call istanbul programatically?

That wasn't possible with jscoverage and that's the reason why we shelled out with it.

sam-falvo replied:
Maybe I'm really misunderstanding the documentation, but I didn't find APIs for Istanbul which replicated the shell behavior.  Maybe they exist, but if so, I'd argue that they need a lot better documentation.

Can we keep the shell-out behavior for now and consider programmatically using Istanbul as a separate PR?  I'd like to get Whiskey back in the hands of people to use.

Kami replied:
https://github.com/yahoo/istanbul/blob/master/lib/command/instrument.js#L95

Yeah, their API for doing instrumentation programatically is not the best.

Commit:
Finishing touches and code clean-up before final review

* Reverted hard-wired library dependencies on lib-cov.  Correct behavior
  has been observationally confirmed, so it's time to wrap things up and
  put things back in order.

  As a feature-request, rch requested that we integrate Rewire into
  Whiskey to better support this kind of testing, where we want to use
  the examples with code-coverage.

* Fixed many missing "global variable" leaks.

* Fixed style issues, such as " vs ' for strings.

* Removed unused references to fs module and fs.appendFileSync().

* I believe I did the right thing to list Istanbul as a proper npm
  dependency for Whiskey.

Commit:
Removing unnecessary code

Commit:
Debugging -- why u no work all of a sudden?!?!

Commit:
Missing fs module in coverage.js

Commit:
removing debugging hardwire to lib-cov

sam-falvo commented:
Code complete.
